### ISSUE TITLE
Bug in reading in Operational Template

### ISSUE THREAD

**Issue by wolandscat** [openEHR/archie#17]
@pieterbos There is a bug that we found the other week while I was at Marand, but can no longer locate in the source which is where the parser wrongly instantiates component_terminologies from an OPT due to a simple error in the way the ODIN is converted to JSON and then instantiated. It is an error I made myself in the original Eiffel code, and the solution is a helper class that has a single attribute, component_terminologies. 

If this error is fixed, it will allow correct reparsing of an OPT.

**Comment by wolandscat** [comment]
We probably do need to leave the specialisation section in the OPT, since it makes OPT parsing more like normal archetype parsing, and the problem with having no specialisation section is that OPTs will always have specialised codes, but with no parent to indicate where to get the parent codes from.

This however is not the main problem with the roundtrip - the component_terminologies is, but it's easy to fix (I just can't find the relevant code...)

**Comment by pieterbos** [comment]
Hello Thomas,

Together with @t-de-jong I just added a branch that adds a roundtrip ADL test for operational templates. It creates an operational template, serializes it and parses it again. To make it pass, I had to add the specialized section to the grammar.

It parses component terminologies just fine. Maybe there is a difference in syntax in what you tested at Marand and what we use? Archie outputs the following syntax:

```
component_terminologies
    < 
        ["openEHR-EHR-OBSERVATION.ovl-blood_pressure-blood_pressure-001.v1.0.0"] = <
            term_definitions = <
                ["en"] = <
                    ["id1"] = <
                        text = <"Blood Pressure">
                        description = <"The local measurement of arterial blood pressure which is a surrogate for arterial. pressure in the systemic circulation.  Most commonly, use of the term 'blood pressure' refers to measurement of brachial artery pressure in the upper arm.">
                    >
                    ["id2"] = <
                        text = <"history">
                        description = <"History Structural node.">
                    >
                    ["id5"] = <
                        text = <"Systolic">
                        description = <"Peak systemic arterial blood pressure  - measured in systolic or contraction phase of the heart cycle.">
                    >
```

**Comment by pieterbos** [comment]
Oh, and the code that parses the operational template is here:

https://github.com/openEHR/archie/blob/master/aom/src/main/java/com/nedap/archie/adlparser/treewalkers/ADLListener.java#L176

No helper object is needed with this approach.

**Comment by wolandscat** [comment]
When we ran this the other week, this line failed:

```
template.setComponentTerminologies(OdinObjectParser.convert(ctx.odin_text(), mapType));
```

Unfortunately I did not take a screen snapshot. I'll get the Marand guys to try to reproduce it and note it here.

**Comment by pieterbos** [comment]
The problem was a difference in component_terminologies syntax. Fixed in https://github.com/openEHR/archie/pull/20 to accept both, and to generate the same as the ADL workbench.

**Comment by pieterbos** [comment]
Issue has been solved in #20
